### PR TITLE
#844: Implement equals and hashCode for MapEnvelope

### PR DESCRIPTION
**Pull request by proshin-roman**
Pull request for #844.

- implemented methods `equals` and `hashCode` for `MapEnvelope` class. Implementation depends on the entries of the map.
- added new test cases to cover implemented methods
- added a new puzzle to add equals/hashCode for CollectionEnvelope

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#865](https://github.com/yegor256/cactoos/issues/865) is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/C63314D6Z/1361153494); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @llorllale[/z](https://www.0crat.com/u/llorllale) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer

File: src/test/java/org/cactoos/map/MapEnvelopeTest.java [hunk 1]
```diff
@@ -203,4 +207,103 @@ public void mapContainsValueFalse() {
...
+                new IsEqual<>("Totally different type")
+            )
+        );
+    }
+
+    @Test
+    public void mapEqualsToMapWithSameEntries() {
+        final MapEntry<String, String> entry =
+            new MapEntry<>("key2", "value2");
+        MatcherAssert.assertThat(
```
Comment (paulodamaso) [inline src/test/java/org/cactoos/map/MapEnvelopeTest.java]: @proshin-roman To enforce value comparison I'd create a new instance of `MapEntry` with the same key / value pair `("key2", "value2")` and use it to check if two maps with the same key / value are equals. What do you think?

File: src/test/java/org/cactoos/map/MapEnvelopeTest.java [hunk 2]
```diff
@@ -203,4 +207,103 @@ public void mapContainsValueFalse() {
...
+
+    @Test
+    public void hashCodeDependsOnItems() {
+        final MapEntry<String, String> entry =
+            new MapEntry<>("key9", "value9");
+        MatcherAssert.assertThat(
+            "hashCode returns different results for same entries",
+            new MapOf<String, String>(entry).hashCode(),
+            new IsEqual<>(new MapOf<String, String>(entry).hashCode())
+        );
```
Comment (paulodamaso) [inline src/test/java/org/cactoos/map/MapEnvelopeTest.java]: @proshin-roman Same as above here, just to enforce that we are using two different objects I think we should create another instance of `MapEntry` with the same values.

File: src/test/java/org/cactoos/map/MapEnvelopeTest.java (hunk 1 above)
Comment (proshin-roman) [reply src/test/java/org/cactoos/map/MapEnvelopeTest.java]: @paulodamaso Well, ok, it makes sense, I will fix it.

**Comment by proshin-roman** [comment]
@paulodamaso I've fixed all comments - please review the PR again

**Comment by paulodamaso** [comment]
@rultor merge please

**Comment by llorllale** [comment]
@rultor merge

**Comment by 0crat** [comment]
@elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova) please review this job completed by @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#865](https://github.com/yegor256/cactoos/issues/865) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @llorllale[/z](https://www.0crat.com/u/llorllale)

**Comment by elenavolokhova** [comment]
@paulodamaso According to our Policy:
> The code reviewer found at least three problems in the code (discuss);
Only two issues were found here.

Please, confirm that you will try to find at least 3 issues next time.

**Comment by elenavolokhova** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 point(s) just awarded to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova)

**Pull request merged**